File: backend/internal/repositories/personrepository.go
package repositories

import (
	"keybook/backend/internal/dtos"

	"github.com/pocketbase/dbx"
	"github.com/pocketbase/pocketbase/core"
	"github.com/pocketbase/pocketbase/models"
)

type IPersonRepository interface {
	GetPersonById(personId string) (dtos.PersonDto, error)
	CreatePerson(name, DOB, userID string) (dtos.PersonDto, error)
	UpdatePerson(id, name, DOB string) error
	DeletePerson(id string) error
}

type PersonRepository struct {
	app core.App
}

func NewPersonRepository(app core.App) IPersonRepository {
	return &PersonRepository{app}
}

func (r *PersonRepository) GetPersonById(personId string) (dtos.PersonDto, error) {
	query := r.app.Dao().DB().Select(
		"p.id",
		"p.name",
		"p.DOB",
		"p.user",
		"p.profileImage",
	).From(
		"persons p",
	).Where(
		dbx.NewExp("p.id = {:personId}", dbx.Params{"personId": personId}),
	)

	var result dtos.PersonDto
	return result, query.One(&result)
}

func (r *PersonRepository) CreatePerson(name, DOB, userID string) (dtos.PersonDto, error) {
	col, err := r.app.Dao().FindCollectionByNameOrId("persons")
	if err != nil {
		return dtos.PersonDto{}, err
	}
	record := models.NewRecord(col)
	record.Set("name", name)
	record.Set("DOB", DOB)
	if userID != "" {
		record.Set("user", userID)
	}
	if err := r.app.Dao().SaveRecord(record); err != nil {
		return dtos.PersonDto{}, err
	}
	return dtos.PersonDto{
		Id:   record.GetId(),
		Name: record.GetString("name"),
		DOB:  record.GetString("DOB"),
		User: record.GetString("user"),
	}, nil
}

func (r *PersonRepository) UpdatePerson(id, name, DOB string) error {
	record, err := r.app.Dao().FindRecordById("persons", id)
	if err != nil {
		return err
	}
	record.Set("name", name)
	record.Set("DOB", DOB)
	return r.app.Dao().SaveRecord(record)
}

func (r *PersonRepository) DeletePerson(id string) error {
	record, err := r.app.Dao().FindRecordById("persons", id)
	if err != nil {
		return err
	}
	return r.app.Dao().DeleteRecord(record)
}

File: frontend/src/lib/repositories/person/person-repository.ts
import PocketBase from 'pocketbase';
import type { IPersonModel } from '$lib/models/person-models';

export interface IPersonRepository {
	getAllPersons(): Promise<IPersonModel[]>;
	getPersonById(id: string): Promise<IPersonModel>;
	getPersonByUserId(userId: string): Promise<IPersonModel | null>;
	createPerson(name: string, dob: string, userId?: string): Promise<IPersonModel>;
	updatePerson(id: string, name: string, dob: string): Promise<IPersonModel>;
	deletePerson(id: string): Promise<void>;
}

export class PersonRepository implements IPersonRepository {
	constructor(private readonly backendClient: PocketBase) {}

	async getAllPersons(): Promise<IPersonModel[]> {
		return await this.backendClient.collection('persons').getFullList<IPersonModel>();
	}

	async getPersonById(id: string): Promise<IPersonModel> {
		const allPersons = await this.backendClient.collection('persons').getFullList<IPersonModel>();
		const person = allPersons.find((candidatePerson) => candidatePerson.id === id);
		if (!person) throw new Error(`Person not found: ${id}`);
		return person;
	}

	async getPersonByUserId(userId: string): Promise<IPersonModel | null> {
		const allPersons = await this.backendClient.collection('persons').getFullList<IPersonModel>();
		return allPersons.find((person) => person.user === userId) ?? null;
	}

	async createPerson(name: string, dob: string, userId?: string): Promise<IPersonModel> {
		const payload: Record<string, string> = { name, DOB: dob };
		if (userId) payload.user = userId;
		return await this.backendClient.collection('persons').create<IPersonModel>(payload);
	}

	async updatePerson(id: string, name: string, dob: string): Promise<IPersonModel> {
		return await this.backendClient
			.collection('persons')
			.update<IPersonModel>(id, { name, DOB: dob });
	}

	async deletePerson(id: string): Promise<void> {
		await this.backendClient.collection('persons').delete(id);
	}
}
